dds/idl/marshal_generator.cpp

```diff
@@ -1076,53 +1406,120 @@ namespace {
           << "_array(arr" << accessor << suffix << ", " << n_elems << ");\n";
       } else { // Enum, String, Struct, Array, Sequence, Union
-        {
-          string indent = "  ";
-          NestedForLoops nfl("CORBA::ULong", "i", arr, indent);
-          if (!use_cxx11 && (elem_cls & CL_ARRAY)) {
-            const string typedefname = scoped(arr->base_type()->name());
+        string indent = "  ";
+        string suffix = "";
+        NestedForLoops nfl("CORBA::ULong", "i", arr, indent);
+
+        if (!use_cxx11 && (elem_cls & CL_ARRAY)) {
+          const string typedefname = scoped(arr->base_type()->name());
+          be_global->impl_ <<
+            indent << typedefname << "_var tmp = " << typedefname << "_alloc();\n" <<
+            indent << typedefname << "_forany fa = tmp.inout();\n" <<
+            indent <<  "    if (!(strm >> fa)) {\n";
+        } else {
+          string intro;
+          be_global->impl_ <<
+            "    if (!" <<
+              streamCommon("", arr->base_type(), string(">> ") + "arr" + nfl.index_, intro) <<
+              ") {\n";
+        }
+        if (try_construct == tryconstructfailaction_use_default) {
+          if (elem_cls & CL_ARRAY) {
             be_global->impl_ <<
-              indent << typedefname << "_var tmp = " << typedefname
-              << "_alloc();\n" <<
-              indent << typedefname << "_forany fa = tmp.inout();\n"
-              << streamAndCheck(">> fa", indent.size()) <<
-              indent << typedefname << "_copy(arr" << nfl.index_ <<
-              ", tmp.in());\n";
+              "      " << type_to_default(elem, "arr" + nfl.index_) <<
+              "      strm.set_construction_status(Serializer::ConstructionSuccessful);\n";
           } else {
-            string suffix = (elem_cls & CL_STRING) ? (use_cxx11 ? "" : ".out()") : "";
-            string pre;
-            if (use_cxx11 && (elem_cls & (CL_ARRAY | CL_SEQUENCE))) {
-              pre = "IDL::DistinctType<" + cxx_elem + ", " +
-                dds_generator::scoped_helper(arr->base_type()->name(), "_") + "_tag>(";
-              suffix += ')';
+            be_global->impl_ << "      " << type_to_default(elem, "arr" + nfl.index_ + suffix) <<
+                                "      strm.set_construction_status(Serializer::ConstructionSuccessful);\n";
+          }
+        } else if ((try_construct == tryconstructfailaction_trim) && (elem_cls & CL_BOUNDED) &&
+                   (elem_cls & (CL_STRING | CL_SEQUENCE))) {
+          if (elem_cls & CL_STRING) {
+            string check_not_empty = use_cxx11 ? "!arr" + nfl.index_ + ".empty()" : "arr" + nfl.index_ + ".in()";
+            string get_length = use_cxx11 ? "arr" + nfl.index_ + ".length()" : "ACE_OS::strlen(arr" + nfl.index_ + ".in())";
+            string inout = use_cxx11 ? "" : ".inout()";
+            if (use_cxx11){
+              be_global->impl_ <<
+                "        if (strm.good_bit() && " << check_not_empty << " && ("
```

Comment:
This is the same issue as my other comment. `good_bit` isn't fit to be used in generated code.